classes/quiz_settings.php

```diff
@@ -147,6 +155,185 @@ protected static function define_properties() {
                 'default' => '',
                 'null' => NULL_ALLOWED,
             ],
+            'config' => [
+                'type' => PARAM_RAW,
+                'default' => '',
+                'null' => NULL_ALLOWED,
+            ],
+        ];
+    }
+
+    /**
+     * Hook to execute before an update.
+     *
+     * Please note that at this stage the data has already been validated and therefore
+     * any new data being set will not be validated before it is sent to the database.
+     */
+    protected function before_update() {
+        $this->before_save();
+    }
+
+    /**
+     * Hook to execute before a create.
+     *
+     * Please note that at this stage the data has already been validated and therefore
+     * any new data being set will not be validated before it is sent to the database.
+     */
+    protected function before_create() {
+        $this->before_save();
+    }
+
+    /**
+     * As there is no hook for before both create and update, this function is called by both hooks.
+     *
+     * @throws \CFPropertyList\IOException
+     * @throws \CFPropertyList\PListException
+     * @throws \DOMException
+     * @throws \coding_exception
+     */
+    private function before_save() {
+        // Recalculate config and config key.
+        $this->compute_config();
+        $this->compute_config_key();
+    }
+
+    /**
+     * Generate the config key from the config string.
+     *
+     * @throws \CFPropertyList\IOException
+     * @throws \CFPropertyList\PListException
+     * @throws \DOMException
+     * @throws \coding_exception
+     */
+    private function compute_config_key() {
+        $config = $this->get('config');
+        $configkey = config_key::generate($config)->get_hash();
+        $this->set('configkey', $configkey);
+    }
+
+    /**
+     * Create or update the config string based on the current quiz settings.
+     *
+     * @throws \CFPropertyList\IOException
+     * @throws \CFPropertyList\PListException
+     * @throws \DOMException
+     * @throws \coding_exception
+     */
+    private function compute_config() {
+        $settings = $this->to_record();
+        // Get existing config.
+        $config = $this->get('config');
+        // Parse basic settings into a property list.
+        $this->plist = new property_list($config);
```

Comment:
Awesome, but as it's not in the constructor, I'd wrap it to $this->get_plist() and initialise there or move to constructor if possible. Otherwise we need to check everytime that it's set. Sorry should point it out first go

Reply:
Of course. Moved to constructor